file-and-cloudinary-upload/app/routes/cloudinary-upload.tsx

```diff
@@ -26,10 +26,10 @@ export const action = async ({ request }: ActionArgs) => {
   const imgSrc = formData.get("img");
   const imgDesc = formData.get("desc");
   if (!imgSrc) {
-    return json({ error: "something wrong" });
+    return json({ error: "something wrong", imgDesc: null, imgSrc: null });
   }
 
-  return json({ imgDesc, imgSrc });
+  return json({ error: null, imgDesc, imgSrc });
```

Comment:
Isn't our type inference smart enough to detect the different return shapes?

Reply:
No, we need to add these if we don't want to do extra checks in the component for existence

Without this we would need to do

```ts
const actionData = useActionData<typeof action>();

if('error' in actionData && actionData.error === "whatever") {
  // ...
}
```

With this change we can just do

```ts
const actionData = useActionData<typeof action>();

if(actionData.?error === "whatever") {
  // ...
}
```